SYCL/ESIMD/spec_const/Inputs/spec_const_common.hpp

```diff
@@ -0,0 +1,89 @@
+//==--------------- spec_const_common.h  - DPC++ ESIMD on-device test -----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// The test checks that ESIMD kernels support specialization constants for all
+// basic types, particularly a specialization constant can be redifined and
+// correct new value is used after redefinition.
+
+#include "esimd_test_utils.hpp"
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/esimd.hpp>
+
+#include <iostream>
+#include <vector>
+
+using namespace cl::sycl;
+
+template <typename AccessorTy>
+ESIMD_INLINE void do_store(AccessorTy acc, int i, spec_const_t val) {
+  using namespace sycl::INTEL::gpu;
+#if STORE == 0
+  // bool
+  scalar_store(acc, i, val ? 1 : 0);
+#elif STORE == 1
+  // block
+  block_store(acc, i, simd<spec_const_t, 2>{val});
+#elif STORE == 2
+  // scalar
+  scalar_store(acc, i, val);
+#endif
+}
+
+class ConstID;
+class TestKernel;
+
+int main(int argc, char **argv) {
+  queue q(esimd_test::ESIMDSelector{}, esimd_test::createExceptionHandler());
+
+  auto dev = q.get_device();
+  std::cout << "Running on " << dev.get_info<info::device::name>() << "\n";
+
+  const int n_times = 2;
+  std::vector<container_t> output(n_times);
+  std::vector<container_t> etalon = {DEF_VAL, REDEF_VAL};
+
+  bool passed = true;
+  for (int i = 0; i < n_times; i++) {
+    sycl::program prg(q.get_context());
+
+    auto spec_const = prg.set_spec_constant<ConstID>((spec_const_t)DEF_VAL);
+    if (i % 2 != 0)
+      spec_const = prg.set_spec_constant<ConstID>((spec_const_t)REDEF_VAL);
```

Comment:
```suggestion
    auto spec_const = prg.set_spec_constant<ConstID>(static_cast<spec_const_t>(etalon[i]));
```

Reply:
The idea here is to check that already initialized constant can be overwritten.

According to [standards proposals](https://github.com/codeplaysoftware/standards-proposals/blob/master/spec-constant/index.md):
> A cl::sycl::experimental::spec_constant object is considered initialized once the result of a cl::sycl::program::set_spec_constant is assigned to it.
> A specialization constant value can be overwritten if the program was not built before by recalling set_spec_constant with the same ID and the new value. Although the type T of the specialization constant must remain the same.

I'll add a comment to clarify this part.